base/math.jl

```diff
@@ -166,6 +170,43 @@ function hypot{T<:AbstractFloat}(x::T, y::T)
     x * sqrt(one(r)+r*r)
 end
 
+# 3-arg hypot
+hypot(x::Real, y::Real, z::Real) =
+    hypot(promote(float(x), float(y), float(z))...)
+function hypot{T<:AbstractFloat}(x::T, y::T, z::T)
+    x = abs(x)
+    y = abs(y)
+    z = abs(z)
+    m = max(x, y, z)
+    if isinf(m)
+        return m
+    elseif isfinite(3*m*m) && m*m != 0 # Scaling not necessary
+        return sqrt(abs2(x) + abs2(y) + abs2(z))
+    elseif m == 0
+        o = one(m)
+        return m*sqrt(abs2(x/o) + abs2(y/o) + abs2(z/o))
+    else
+        return m*sqrt(abs2(x/m) + abs2(y/m) + abs2(z/m))
```

Comment:
Square roots can usually be vectorized. I'm also assuming that something like `@fastmath` is used to ensure that `sqrt` doesn't throw exceptions, and to avoid calling `isinf`.

Anyway, this function begins to look so complex that it seems unlikely to be inlined and vectorized.